#! /usr/bin/env python

PACKAGE='pgr_camera'
import roslib; roslib.load_manifest(PACKAGE)

from driver_base.msg import SensorLevels
from dynamic_reconfigure.parameter_generator import *

gen = ParameterGenerator()

# @todo Think about levels. Setting most of these to STOP to guarantee atomicity.
# @todo Double check these ranges, defaults
mode_enum = gen.enum( [ gen.const("StreamingMode", str_t, "streaming", "Streaming mode"),
                        gen.const("PolledMode", str_t, "polled", "Polled mode") ], "Enum to set the mode.")
format_enum = gen.enum( [ gen.const("Y8", str_t, "Y8", "Y8 mode"),
                          gen.const("RGB", str_t, "RGB", "RGB mode") ], "Enum to set the format.")
#       Name                 Type      Reconfiguration level             Description   Default   Min   Max
gen.add("trigger_mode",      str_t,    SensorLevels.RECONFIGURE_STOP,    "Camera mode (streaming vs. polled)", "streaming", edit_method = mode_enum)
gen.add("auto_exposure",     bool_t,   SensorLevels.RECONFIGURE_RUNNING, "Sets the camera exposure duration to automatic. Causes the @b exposure setting to be ignored.", True)
gen.add("exposure",          double_t, SensorLevels.RECONFIGURE_RUNNING, "Camera exposure time in seconds.", 0.015, 0, 1)
gen.add("auto_gain",         bool_t,   SensorLevels.RECONFIGURE_RUNNING, "Sets the analog gain to automatic. Causes the @b gain setting to be ignored.", True)
gen.add("gain",              int_t,    SensorLevels.RECONFIGURE_RUNNING, "The gain level in dB.", 0, 0, 16)
gen.add("auto_whitebalance", bool_t,   SensorLevels.RECONFIGURE_RUNNING, "Whether whitebalance will continuously adjust to the current scene. Causes the @b whitebalance_red and @b whitebalance_blue settings to be ignored.", True)
gen.add("whitebalance_red",  int_t,    SensorLevels.RECONFIGURE_RUNNING, "Red gain as a percentage of the camera default setting.", 0, 0, 100)
gen.add("whitebalance_blue", int_t,    SensorLevels.RECONFIGURE_RUNNING, "Blue gain as a percentage of the camera default setting.", 0, 0, 100)
# @todo ROI, binning...
gen.add("frame_id",          str_t,    SensorLevels.RECONFIGURE_RUNNING, "The optical camera TF frame set in message headers.", "camera")
gen.add("roi_x_offset",          int_t,    SensorLevels.RECONFIGURE_RUNNING, "X offset of the region of interest.", 0, 0, 4000)
gen.add("roi_y_offset",          int_t,    SensorLevels.RECONFIGURE_RUNNING, "Y offset of the region of interest.", 0, 0, 4000)
gen.add("roi_width",             int_t,    SensorLevels.RECONFIGURE_RUNNING, "Width of the region of interest (0 for automatic).", 0, 0, 4000)
gen.add("roi_height",            int_t,    SensorLevels.RECONFIGURE_RUNNING, "Height of the region of interest (0 for automatic).", 0, 0, 4000)
# @todo Add more modes and framerates
gen.add("width",             int_t,    SensorLevels.RECONFIGURE_RUNNING, "Capture width", 640, 640, 1296)
gen.add("height",            int_t,    SensorLevels.RECONFIGURE_RUNNING, "Capture height", 480, 480, 964)
gen.add("format",            str_t,    SensorLevels.RECONFIGURE_RUNNING, "Capture format (Y8,RGB)", "Y8", edit_method = format_enum)
gen.add("frame_rate",        double_t, SensorLevels.RECONFIGURE_RUNNING, "Camera frame rate in Hz.", 30.0, 1.151, 60.861)
gen.add("intrinsics_ini",            str_t,    SensorLevels.RECONFIGURE_RUNNING, "Camera intrinsics INI filename", "test_intrinsics.ini")
gen.add("camera_name",            str_t,    SensorLevels.RECONFIGURE_RUNNING, "Camera Name (used in loading intrinsics INI)", "pgr_camera")

exit(gen.generate(PACKAGE, "pgr_camera_node", "PGRCamera"))
